R/segment_rule.R

```diff
@@ -0,0 +1,283 @@
+
+      obj <-  year_change(object)
+    }
+    obj
+  }
+  #/end date formatting
+  #if subject is one of 'daterange' the object must be changed in the definition to the correct format
+  if (subject %in% c('daterangehour', 'daterangeday', 'daterangeweek', 'daterangemonth', 'daterangequarter', 'daterangeyear')){
+    verb_class <- lookup_verb_class(verb, verbs)
+    object <- daterange_change(subject, object, verb_class)
+  }
+
+  #Define the val func to be an attr or event
+  define_val_func <- function(adj){
+    if (adj == 'variables/'){
+      'attr'
+    } else {
+      'event'
+    }
+  }
+  lookup_verb_arg <- function(verb, verbs) {
+    arg_out <- verbs$arg[verbs$verb == verb]
+    if (length(arg_out) == 0) {
+      stop("'", verb, "' is not a valid API verb")
+    }
+    arg_out
+  }
+
+  create_rule <- function(verb, subject, object, adj, is_distinct){
+    ## define the base argument of the verb function
+    arg <- lookup_verb_arg(verb, verbs)
+    ## get the val_func value
+    val_func <- define_val_func(adj)
+    ## Change to NULL for easier list management
+    if (is.na(arg)) {
+      arg <- NULL
+      object <- NULL
+    }
+    ## Create the top-level list of list-names
+    if (val_func == 'event' && is.null(arg) ) {
+      listnames <- purrr::compact(c('func', arg, 'evt'))
+    } else {
+      listnames <- purrr::compact(c('func', arg, 'val'))
+    }
+    ## Build the list values
+    ## included the first if bucket due to an error with null arg
+    if (!is.null(arg)) {
+      if (val_func == 'event' && arg == 'num') {
+        rule <- purrr::compact(list(verb, object, list(func = 'total', evt = list(val_func,  glue::glue('{adj}{subject}')))))
+      } else if (val_func == 'attr' && arg == 'num' && is_distinct) {
+        rule <- purrr::compact(list(verb, object, list(func = 'total-distinct', field = list(val_func,  glue::glue('{adj}{subject}')))))
+      } else {
+        rule <- purrr::compact(list(verb, object, list(val_func,  glue::glue('{adj}{subject}'))))
+      }
+    } else {
+      rule <- purrr::compact(list(verb, object, list(val_func,  glue::glue('{adj}{subject}'))))
+    }
+    ## Name the list values
+    names(rule) <- listnames
+    #name the sublist values
+    listsubnames <- c("func", "name")
+    if (!is.null(rule$val$func)) {
+      if (rule$val$func == 'total') {
+        names(rule$val$evt) <- listsubnames
+      } else {
+        names(rule$val$field) <- listsubnames
+      }
+    } else if (is.null(arg) && val_func == 'event') {
+      names(rule[['evt']]) <- listsubnames
+    } else {
+      names(rule[['val']]) <- listsubnames
+    }
+    rule
+  }
+
+  #This forms the rule or rule
+  prerule <- create_rule(verb, subject, object, adj, is_distinct)
+
+  #add in the description
+  if(!is.null(description)){
+    prerule$description = description
```

Comment:
We should continue using `<-` instead of switching to `=` here for assignment. Applies to this whole block below.

```suggestion
    prerule$description <- description
```